File: src/edu/uf/interactable/PneumocyteII.java
package edu.uf.interactable;

import edu.uf.compartments.Voxel;
import edu.uf.interactable.Afumigatus.Afumigatus;
import edu.uf.intracellularState.IntracellularModel;
import edu.uf.primitives.Interactions;
import edu.uf.utils.Constants;
import edu.uf.utils.Id;

public class PneumocyteII extends Cell {
    public static final String NAME = "Pneumocyte";

    private static int totalCells = 0;
    
    private int iteration;
    
    /*public static final int ACTIVE = Phenotype.createPhenotype();
    public static final int MIX_ACTIVE = Phenotype.createPhenotype();*/
    
    
    private static int interactionId = Id.getMoleculeId();

    public PneumocyteII(IntracellularModel network) {
        super(network);
        this.iteration = 0;
        PneumocyteII.totalCells = PneumocyteII.totalCells + 1;
    }
    
    public int getInteractionId() {
    	return interactionId;
    }
    
    public boolean isTime() {
		return this.getClock().toc();
	}

    public static int getTotalCells() {
		return totalCells;
	}

	public static void setTotalCells(int totalCells) {
		PneumocyteII.totalCells = totalCells;
	}

	public int getIteration() {
		return iteration;
	}

    
	public void die() {
		if(this.getBooleanNetwork().getState(IntracellularModel.LIFE_STATUS) != Cell.DEAD) {
    		this.getBooleanNetwork().setState(IntracellularModel.LIFE_STATUS, Cell.DEAD);
            PneumocyteII.totalCells--;
        }
    }

    protected boolean templateInteract(Interactable interactable, int x, int y, int z) {
        if (interactable instanceof Afumigatus) 
            return Interactions.typeIIPneumocyteAspergillus(this, (Afumigatus) interactable);
        
        if (interactable instanceof IL6) 
        	return Interactions.secrete(this, (IL6) interactable, Constants.MA_IL6_QTTY, x, y, z, 0); 
        
        if (interactable instanceof TNFa) {
        	Interactions.bind(this, (TNFa) interactable, x, y, z, 0);
            return Interactions.secrete(this, (TNFa) interactable, Constants.MA_TNF_QTTY, x, y, z, 0); 
        }
		
        return interactable.interact(this, x, y, z);
    }

    public void incIronPool(double qtty) {}

    /*public void updateStatus(int x, int y, int z) {
    	super.updateStatus(x, y, z);
    	if(!this.getClock().toc())return;
    	this.processBooleanNetwork();
    	if(this.getBooleanNetwork().hasPhenotype(IntracellularModel.APOPTOTIC))//I STILL DON'T KNOW IF THIS CODE WORKS
    		this.die();
    	//if(this.hasPhenotype(new int[] {Leukocyte.APOPTOTIC})) //I DON'T KNOW IF THIS CODE WORKS
			
    }*/
           
    /**
     * Disabled.
     */
    public void move(Voxel oldVoxel, int steps) {}

	@Override
	public String getName() {
		return NAME;
	}

	@Override
	public int getMaxMoveSteps() {
		// TODO Auto-generated method stub
		return -1;
	}
}

File: cpp/src/edu/uf/interactable/PneumocyteII.cpp
/*
 * PneumocyteII.cpp
 *
 *  Created on: Jul 18, 2024
 *      Author: henriquedeassis
 */

#include "PneumocyteII.h"
#include "TNFa.h"
#include "../primitives/Interactions.h"
#include "../utils/Constants.h"
#include "../utils/Id.h"

namespace edu {
namespace uf {
namespace interactable {

using namespace edu::uf::utils;
using namespace afumigatus;

const int PneumocyteII::classId = Id::getId();
const std::string PneumocyteII::NAME = "Pneumocyte";
int PneumocyteII::totalCells = 0;
int PneumocyteII::interactionId = utils::Id::getId();

PneumocyteII::PneumocyteII(intracellularState::IntracellularModel* network)
    : Cell(network), iteration(0) {
    totalCells++;
}

int PneumocyteII::getClassId() const {
	return classId;
}

int PneumocyteII::getInteractionId() const {
    return interactionId;
}

bool PneumocyteII::isTime() const {
    return this->getClock()->toc();
}

int PneumocyteII::getTotalCells() {
    return totalCells;
}

void PneumocyteII::setTotalCells(int totalCells) {
    PneumocyteII::totalCells = totalCells;
}

int PneumocyteII::getIteration() const {
    return iteration;
}

void PneumocyteII::die() {
    if (this->getBooleanNetwork()->getState(intracellularState::IntracellularModel::LIFE_STATUS) != intracellularState::IntracellularModel::DEAD) {
        this->getBooleanNetwork()->setState(intracellularState::IntracellularModel::LIFE_STATUS,    intracellularState::IntracellularModel::DEAD);
        totalCells--;
    }
}

bool PneumocyteII::templateInteract(interactable::Interactable* interactable, int x, int y, int z) {
    if (dynamic_cast<TNFa*>(interactable)) {
        primitives::Interactions::bind(this, static_cast<Molecule*>(interactable), x, y, z, 0);
        return primitives::Interactions::secrete(this, static_cast<Molecule*>(interactable), utils::constexprants::MA_TNF_QTTY, x, y, z, 0);
    }

    return interactable->interact(this, x, y, z);
}

void PneumocyteII::incIronPool(double qtty) {
    // No implementation needed as per original code
}

void PneumocyteII::move(int x, int y, int z, int steps) {
    // Disabled as per original code
}

std::string PneumocyteII::getName() const {
    return NAME;
}

int PneumocyteII::getMaxMoveSteps() {
    // TODO: Implement as needed
    return -1;
}

} // namespace interactable
} // namespace uf
} // namespace edu
